### PR TITLE
[BEAM-3516] Spanner BatchFn does not respect mutation limits

### PR DESCRIPTION
**Pull request by mairbek**
Estimate the number of mutations in in a group by counting the columns and associated indexes

Rebased https://github.com/apache/beam/pull/4860

### CODE REVIEW

**Comment by mairbek** [comment]
R: @jkff PTAL

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/MutationCellEstimator.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.spanner;
+
+import com.google.cloud.spanner.KeySet;
+import com.google.cloud.spanner.Mutation;
+import com.google.cloud.spanner.Mutation.Op;
+import com.google.common.collect.Iterables;
+
+final class MutationCellEstimator {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/MutationCellEstimator.java]: MutationCellCounter maybe?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java [hunk 2]
```diff
@@ -653,10 +654,14 @@ public CreateTransaction withTimestampBound(TimestampBound timestampBound) {
   @AutoValue
   public abstract static class Write extends PTransform<PCollection<Mutation>, PDone> {
 
+    private long maxNumMutations;
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java]: This variable should not be needed - this is an AutoValue class

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java [hunk 3]
```diff
@@ -19,68 +19,105 @@
...
-      keyParts.put(table.toLowerCase(), KeyPart.create(column.toLowerCase(), desc));
+      keyPartsBuilder().put(table.toLowerCase(), KeyPart.create(column.toLowerCase(), desc));
       return this;
     }
 
-    public SpannerSchema build() {
-      return new SpannerSchema(columns, keyParts);
-    }
-  }
+    abstract SpannerSchema autoBuild();
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java]: Is this one of the supported builder function names for AutoValue? I thought it's supposed to be called build().

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java [hunk 4]
```diff
@@ -19,68 +19,105 @@
...
+      cellsMutatedPerRowBuilder().putAll(Maps.transformValues(
+          cellsMutatedPerColumn().rowMap(),
+          entry -> entry.values().stream().mapToLong(Long::longValue).sum()));
 
-  private SpannerSchema(ArrayListMultimap<String, Column> columns,
-      ArrayListMultimap<String, KeyPart> keyParts) {
-    this.columns = columns;
-    this.keyParts = keyParts;
-    tables = new ArrayList<>(columns.keySet());
+      tablesBuilder().addAll(columns().keySet());
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java]: This makes the method non-idempotent. Is there a way to refactor this to be simpler, and not have the builder method mutate the value?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/MutationCellEstimator.java (hunk 1 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/MutationCellEstimator.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java (hunk 2 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java (hunk 3 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java]: Yes, it's sometimes it is useful to tweak the logic `build()` method. See https://github.com/google/auto/blob/master/value/userguide/builders-howto.md#validate

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java (hunk 4 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java]: We can just use a setter here. Updated.

**Comment by mairbek** [comment]
Rebased

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/MutationUtils.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package org.apache.beam.sdk.io.gcp.spanner;
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/MutationUtils.java]: Please add license headers to make `./gradlew :rat` happy. Alternatively move `isPointDelete` into `MutationCellCounter` - as it's only used from there.

Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/MutationUtils.java]: Done

**Comment by mairbek** [comment]
Run Java PostCommit

**Comment by mairbek** [comment]
Run Java PostCommit

**Pull request merged**